File: go/pkg/lifecycle/lifecycle.go
package lifecycle

import (
	"fmt"
	"time"

	"kclone-go/pkg/models"
	"kclone-go/pkg/scheduler"
	"kclone-go/pkg/state"
)

// CreatePod creates a pod and schedules it
func CreatePod(cs *state.ClusterState, spec models.PodSpec) *models.Pod {
	pod := cs.AddPod(spec)
	scheduler.SchedulePendingPods(cs)
	cs.RefreshServiceEndpoints()
	return pod
}

// DeletePod removes a pod
func DeletePod(cs *state.ClusterState, uid string) error {
	err := cs.RemovePod(uid)
	if err != nil {
		return err
	}
	cs.RefreshServiceEndpoints()
	return nil
}

// HealthCheckPod simulates a health check on a pod
func HealthCheckPod(cs *state.ClusterState, uid string) error {
	pods := cs.GetPods()
	for _, pod := range pods {
		if pod.UID == uid {
			if pod.Status.Phase == "Running" {
				pod.Status.Healthy = true
				pod.Status.StartTime = time.Now().Format(time.RFC3339)
				return nil
			}
			return fmt.Errorf("pod %s is not running", uid)
		}
	}
	return fmt.Errorf("pod %s not found", uid)
}

// RestartPod restarts a failed pod
func RestartPod(cs *state.ClusterState, uid string) error {
	pods := cs.GetPods()
	for _, pod := range pods {
		if pod.UID == uid {
			if pod.Spec.RestartPolicy == "Never" {
				return fmt.Errorf("pod %s has restart policy Never", uid)
			}
			
			// Remove and recreate
			DeletePod(cs, uid)
			newPod := cs.AddPod(pod.Spec)
			newPod.Status.RestartCount = pod.Status.RestartCount + 1
			scheduler.SchedulePendingPods(cs)
			cs.RefreshServiceEndpoints()
			return nil
		}
	}
	return fmt.Errorf("pod %s not found", uid)
}

File: python/src/kclone/lifecycle.py
from __future__ import annotations

from datetime import datetime
from typing import Dict

from .models import PodSpec
from .scheduler import schedule_pending_pods
from .state import ClusterState


def create_pod(
    state: ClusterState,
    name: str,
    image: str,
    cpu_request: int,
    mem_request: int,
    labels: Dict[str, str],
) -> None:
    spec = PodSpec(name=name, image=image, cpu_request=cpu_request, mem_request=mem_request, labels=labels)
    pod = state.add_pod(spec)
    # trigger scheduling attempt
    schedule_pending_pods(state)
    state.refresh_service_endpoints()
    return pod


def delete_pod(state: ClusterState, uid: str) -> None:
    state.remove_pod(uid)
    state.refresh_service_endpoints()


def health_check_pod(state: ClusterState, uid: str) -> bool:
    """Perform a basic health check simulation for a pod."""
    pod = state.get_pod(uid)
    if not pod:
        raise KeyError(f"Pod {uid} not found")
    if pod.status.phase == "Running":
        pod.status.healthy = True
        pod.status.start_time = datetime.now().timestamp()
        return True
    return False


def restart_pod(state: ClusterState, uid: str) -> None:
    """Restart a pod according to its restart policy."""
    pod = state.get_pod(uid)
    if not pod:
        raise KeyError(f"Pod {uid} not found")

    if pod.spec.restart_policy == "Never":
        raise ValueError(f"Pod {uid} has restart policy Never")

    spec = pod.spec
    restart_count = pod.status.restart_count

    # delete and create new
    delete_pod(state, uid)
    new_pod = state.add_pod(spec)
    new_pod.status.restart_count = restart_count + 1
    schedule_pending_pods(state)
    state.refresh_service_endpoints()
